File: VIDEO/_retired_python_video/app/utils/ffmpeg_compat.py
"""FFmpeg 版本兼容：二进制解析、H.264 编码器选择、RTSP 超时参数等。"""
from __future__ import annotations

import os
import subprocess
from typing import List, Optional, Set

# FFmpeg 8+：-stimeout/-rw_timeout 已移除，统一为 -timeout（单位仍为微秒）
_FFMPEG_RTSP_OPEN_TIMEOUT_FLAG: Optional[str] = None
_FFMPEG_SUPPORTS_RW_TIMEOUT: Optional[bool] = None
_RTSP_DEMUXER_HELP: Optional[str] = None
_FFMPEG_ENCODER_NAMES: Optional[Set[str]] = None
_RESOLVED_FFMPEG_BIN: Optional[str] = None


def _is_runnable_binary(path: str) -> bool:
    if not path or not os.path.isfile(path):
        return False
    lower = path.lower()
    if lower.endswith(('.exe', '.cmd', '.bat')):
        return True
    return os.access(path, os.X_OK)


def resolve_ffmpeg_binary() -> str:
    """优先 FFMPEG_PATH；否则 PATH 上的 ffmpeg。"""
    global _RESOLVED_FFMPEG_BIN
    if _RESOLVED_FFMPEG_BIN is not None:
        return _RESOLVED_FFMPEG_BIN
    explicit = (os.getenv('FFMPEG_PATH') or '').strip().strip('"')
    if _is_runnable_binary(explicit):
        _RESOLVED_FFMPEG_BIN = explicit
        return _RESOLVED_FFMPEG_BIN
    _RESOLVED_FFMPEG_BIN = 'ffmpeg'
    return _RESOLVED_FFMPEG_BIN


def ffmpeg_encoder_names() -> Set[str]:
    """探测当前 ffmpeg 可用的编码器名（缓存）。"""
    global _FFMPEG_ENCODER_NAMES
    if _FFMPEG_ENCODER_NAMES is not None:
        return _FFMPEG_ENCODER_NAMES
    names: Set[str] = set()
    try:
        probe = subprocess.run(
            [resolve_ffmpeg_binary(), '-hide_banner', '-encoders'],
            capture_output=True,
            timeout=8,
        )
        text = (probe.stdout or b'').decode(errors='replace')
        for line in text.splitlines():
            # e.g. " V..... libx264              libx264 H.264 ..."
            parts = line.strip().split()
            if len(parts) >= 2 and parts[0].startswith('V'):
                names.add(parts[1])
    except Exception:
        names = set()
    _FFMPEG_ENCODER_NAMES = names
    return _FFMPEG_ENCODER_NAMES


def resolve_view_h264_codec() -> str:
    """预览转推 H.264 编码器：环境变量覆盖，否则 libx264 → h264_nvenc → copy。"""
    forced = (
        os.getenv('VIEW_FFMPEG_CODEC')
        or os.getenv('FFMPEG_VIDEO_CODEC')
        or ''
    ).strip().lower()
    if forced in ('libx264', 'h264_nvenc', 'copy', 'h264_qsv', 'h264_amf'):
        return forced
    encoders = ffmpeg_encoder_names()
    if 'libx264' in encoders:
        return 'libx264'
    if 'h264_nvenc' in encoders:
        return 'h264_nvenc'
    if 'h264_qsv' in encoders:
        return 'h264_qsv'
    if 'h264_amf' in encoders:
        return 'h264_amf'
    return 'copy'


def ffmpeg_option_missing(stderr: bytes, option: str = "") -> bool:
    err = (stderr or b"").decode(errors="replace")
    if "Unrecognized option" in err or "Option not found" in err:
        return True
    if option and f"Option {option} not found" in err:
        return True
    return False


def _rtsp_demuxer_help_text() -> str:
    """读取 ffmpeg RTSP demuxer 帮助，用于判断各版本支持的超时参数。"""
    global _RTSP_DEMUXER_HELP
    if _RTSP_DEMUXER_HELP is not None:
        return _RTSP_DEMUXER_HELP
    try:
        probe = subprocess.run(
            [resolve_ffmpeg_binary(), "-hide_banner", "-h", "demuxer=rtsp"],
            capture_output=True,
            timeout=5,
        )
        _RTSP_DEMUXER_HELP = (probe.stdout or b"").decode(errors="replace")
    except Exception:
        _RTSP_DEMUXER_HELP = ""
    return _RTSP_DEMUXER_HELP


def _rtsp_demuxer_has_option(option: str) -> bool:
    name = option.lstrip("-")
    text = _rtsp_demuxer_help_text()
    return f"-{name}" in text


def ffmpeg_rtsp_open_timeout_flag() -> str:
    """返回当前 ffmpeg 支持的 RTSP 连接超时参数名。"""
    global _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG
    if _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG is not None:
        return _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG
    if _rtsp_demuxer_has_option("stimeout"):
        _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG = "-stimeout"
    else:
        _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG = "-timeout"
    return _FFMPEG_RTSP_OPEN_TIMEOUT_FLAG


def ffmpeg_supports_rw_timeout() -> bool:
    """FFmpeg 8+ 已移除 -rw_timeout，仅保留 -timeout 覆盖 socket I/O。"""
    global _FFMPEG_SUPPORTS_RW_TIMEOUT
    if _FFMPEG_SUPPORTS_RW_TIMEOUT is not None:
        return _FFMPEG_SUPPORTS_RW_TIMEOUT
    if _rtsp_demuxer_has_option("rw_timeout"):
        _FFMPEG_SUPPORTS_RW_TIMEOUT = True
        return _FFMPEG_SUPPORTS_RW_TIMEOUT
    # 帮助文本不可用时，回退到运行时探测（lavfi 轻量输入）
    try:
        probe = subprocess.run(
            [
                resolve_ffmpeg_binary(),
                "-hide_banner",
                "-rw_timeout",
                "1",
                "-f",
                "lavfi",
                "-i",
                "nullsrc=s=1x1:d=0.01",
                "-frames:v",
                "1",
                "-f",
                "null",
                "-",
            ],
            capture_output=True,
            timeout=8,
        )
        _FFMPEG_SUPPORTS_RW_TIMEOUT = not ffmpeg_option_missing(probe.stderr, "rw_timeout")
    except Exception:
        _FFMPEG_SUPPORTS_RW_TIMEOUT = False
    return _FFMPEG_SUPPORTS_RW_TIMEOUT


def ffmpeg_rtsp_timeout_args(open_us: int, io_us: int) -> List[str]:
    """按当前 ffmpeg 版本组装 RTSP 超时参数。"""
    open_flag = ffmpeg_rtsp_open_timeout_flag()
    if ffmpeg_supports_rw_timeout():
        return [open_flag, str(open_us), "-rw_timeout", str(io_us)]
    # FFmpeg 8：-timeout 同时覆盖连接与读写，取较大值
    return [open_flag, str(max(open_us, io_us))]

File: DEVICE/iot-video/iot-video-biz/src/main/java/com/basiclab/iot/video/util/FfmpegCompat.java
package com.basiclab.iot.video.util;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.concurrent.TimeUnit;

/**
 * FFmpeg binary resolution and encoder / RTSP timeout compatibility — aligned with
 * {@code VIDEO/app/utils/ffmpeg_compat.py}.
 */
public final class FfmpegCompat {

    private static volatile String resolvedBinary;
    private static volatile String configuredBinary;
    private static volatile Set<String> encoderNames;
    private static volatile String rtspOpenTimeoutFlag;
    private static volatile Boolean supportsRwTimeout;
    private static volatile String rtspDemuxerHelp;

    private FfmpegCompat() {}

    public static void setConfiguredBinary(String path) {
        configuredBinary = path;
        resolvedBinary = null;
    }

    public static String resolveFfmpegBinary() {
        if (resolvedBinary != null) {
            return resolvedBinary;
        }
        String explicit = (System.getenv("FFMPEG_PATH") != null ? System.getenv("FFMPEG_PATH") : "").trim()
                .replace("\"", "");
        if (explicit.isBlank() && configuredBinary != null && !configuredBinary.isBlank()) {
            explicit = configuredBinary.trim();
        }
        if (isRunnableBinary(explicit)) {
            resolvedBinary = explicit;
            return resolvedBinary;
        }
        resolvedBinary = "ffmpeg";
        return resolvedBinary;
    }

    private static boolean isRunnableBinary(String path) {
        if (path == null || path.isBlank()) {
            return false;
        }
        File file = new File(path);
        if (!file.isFile()) {
            return false;
        }
        String lower = path.toLowerCase(Locale.ROOT);
        if (lower.endsWith(".exe") || lower.endsWith(".cmd") || lower.endsWith(".bat")) {
            return true;
        }
        return file.canExecute();
    }

    public static Set<String> ffmpegEncoderNames() {
        if (encoderNames != null) {
            return encoderNames;
        }
        Set<String> names = new HashSet<>();
        try {
            Process process = new ProcessBuilder(
                    resolveFfmpegBinary(), "-hide_banner", "-encoders"
            ).redirectErrorStream(true).start();
            boolean finished = process.waitFor(8, TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
            }
            String text = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            for (String line : text.split("\n")) {
                String trimmed = line.strip();
                String[] parts = trimmed.split("\\s+");
                if (parts.length >= 2 && parts[0].startsWith("V")) {
                    names.add(parts[1]);
                }
            }
        } catch (Exception ignored) {
            names = new HashSet<>();
        }
        encoderNames = names;
        return encoderNames;
    }

    public static String resolveViewH264Codec() {
        String forced = firstNonBlank(
                System.getenv("VIEW_FFMPEG_CODEC"),
                System.getenv("FFMPEG_VIDEO_CODEC")
        ).toLowerCase(Locale.ROOT);
        if (forced.equals("libx264") || forced.equals("h264_nvenc") || forced.equals("copy")
                || forced.equals("h264_qsv") || forced.equals("h264_amf")) {
            return forced;
        }
        Set<String> encoders = ffmpegEncoderNames();
        if (encoders.contains("libx264")) {
            return "libx264";
        }
        if (encoders.contains("h264_nvenc")) {
            return "h264_nvenc";
        }
        if (encoders.contains("h264_qsv")) {
            return "h264_qsv";
        }
        if (encoders.contains("h264_amf")) {
            return "h264_amf";
        }
        return "copy";
    }

    public static List<String> ffmpegRtspTimeoutArgs(int openUs, int ioUs) {
        String openFlag = ffmpegRtspOpenTimeoutFlag();
        if (ffmpegSupportsRwTimeout()) {
            return List.of(openFlag, String.valueOf(openUs), "-rw_timeout", String.valueOf(ioUs));
        }
        return List.of(openFlag, String.valueOf(Math.max(openUs, ioUs)));
    }

    /** Non-RTSP input timeout args — mirrors {@code camera.py} FFmpegDaemon lines 303–306. */
    public static List<String> ffmpegNonRtspTimeoutArgs(int ioUs) {
        if (ffmpegSupportsRwTimeout()) {
            return List.of("-rw_timeout", String.valueOf(ioUs));
        }
        if ("-timeout".equals(ffmpegRtspOpenTimeoutFlag())) {
            return List.of("-timeout", String.valueOf(ioUs));
        }
        return Collections.emptyList();
    }

    public static String ffmpegRtspOpenTimeoutFlag() {
        if (rtspOpenTimeoutFlag != null) {
            return rtspOpenTimeoutFlag;
        }
        if (rtspDemuxerHasOption("stimeout")) {
            rtspOpenTimeoutFlag = "-stimeout";
        } else {
            rtspOpenTimeoutFlag = "-timeout";
        }
        return rtspOpenTimeoutFlag;
    }

    public static boolean ffmpegSupportsRwTimeout() {
        if (supportsRwTimeout != null) {
            return supportsRwTimeout;
        }
        if (rtspDemuxerHasOption("rw_timeout")) {
            supportsRwTimeout = true;
            return true;
        }
        try {
            Process process = new ProcessBuilder(
                    resolveFfmpegBinary(),
                    "-hide_banner",
                    "-rw_timeout", "1",
                    "-f", "lavfi",
                    "-i", "nullsrc=s=1x1:d=0.01",
                    "-frames:v", "1",
                    "-f", "null",
                    "-"
            ).redirectErrorStream(true).start();
            boolean finished = process.waitFor(8, TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
            }
            String err = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            supportsRwTimeout = !ffmpegOptionMissing(err, "rw_timeout");
        } catch (Exception e) {
            supportsRwTimeout = false;
        }
        return supportsRwTimeout;
    }

    private static boolean rtspDemuxerHasOption(String option) {
        String name = option.startsWith("-") ? option.substring(1) : option;
        return rtspDemuxerHelpText().contains("-" + name);
    }

    private static String rtspDemuxerHelpText() {
        if (rtspDemuxerHelp != null) {
            return rtspDemuxerHelp;
        }
        try {
            Process process = new ProcessBuilder(
                    resolveFfmpegBinary(), "-hide_banner", "-h", "demuxer=rtsp"
            ).redirectErrorStream(true).start();
            boolean finished = process.waitFor(5, TimeUnit.SECONDS);
            if (!finished) {
                process.destroyForcibly();
            }
            rtspDemuxerHelp = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        } catch (Exception e) {
            rtspDemuxerHelp = "";
        }
        return rtspDemuxerHelp;
    }

    private static boolean ffmpegOptionMissing(String stderr, String option) {
        if (stderr == null) {
            return false;
        }
        if (stderr.contains("Unrecognized option") || stderr.contains("Option not found")) {
            return true;
        }
        if (option != null && !option.isBlank()) {
            return stderr.contains("Option " + option + " not found");
        }
        return false;
    }

    private static String firstNonBlank(String... values) {
        for (String value : values) {
            if (value != null && !value.isBlank()) {
                return value.trim();
            }
        }
        return "";
    }

    public static int envInt(String name, int defaultValue) {
        String raw = System.getenv(name);
        if (raw == null || raw.isBlank()) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(raw.trim());
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    public static String envStr(String name, String defaultValue) {
        String raw = System.getenv(name);
        if (raw == null || raw.isBlank()) {
            return defaultValue;
        }
        return raw.trim();
    }
}
